### PR TITLE
Implement optimize call

### PR DESCRIPTION
**Pull request by jimmyhmiller**
This dispatches to a c func for doing the dynamic lookup. I experimented with chain on the proc but wasn't able to detect which call signs would be monomorphic vs polymorphic. There is definitely room for optimization here, but it does reduce exits.

### CODE REVIEW

**Comment by noahgibbs** [comment]
That makes sense. Does it increase the %-in-YJIT for Railsbench at all, or just change which side exits we get?

**Comment by jimmyhmiller** [comment]
@noahgibbs Yeah we don't side exit on `call` at all anymore in railsbench and go from 88.6% to 89.8% When running

```bash
./run_once.sh --yjit --yjit-call-threshold=1 --yjit-stats benchmarks/railsbench/benchmark.rb
```

**Comment by noahgibbs** [comment]
That's not bad. Probably means we're still side-exiting a lot of the time in the Rails router... But that's a perfectly good reduction in side exits!

File: yjit.c [hunk 1]
```diff
@@ -716,6 +716,14 @@ rb_get_iseq_body_param_opt_table(const rb_iseq_t *iseq)
     return iseq->body->param.opt_table;
 }
 
+VALUE
+rb_optimize_call(VALUE *recv, rb_execution_context_t *ec, int argc, VALUE *argv, int kw_splat, VALUE block_handler) {
```
Comment (k0kubun) [inline yjit.c:720]: nitpick: It seems a bit inconsistent with other functions

```suggestion
rb_optimize_call(VALUE *recv, rb_execution_context_t *ec, int argc, VALUE *argv, int kw_splat, VALUE block_handler)
{
```

File: yjit/src/codegen.rs [hunk 2]
```diff
@@ -5557,8 +5557,54 @@ fn gen_send_general(
...
+                        let sp = asm.lea(ctx.sp_opnd(0));
+
+                        // Write interpreter SP into CFP.
+                        // Needed in case the callee yields to the block.
+                        jit_save_pc(jit, asm);
+                        // Store incremented PC into current control frame in case callee raises.
+                        gen_save_sp(jit, asm, ctx);
+
+                        let kw_splat = flags & VM_CALL_KW_SPLAT;
+                        let argv = asm.lea(Opnd::mem(64, sp, -(argc) * SIZEOF_VALUE_I32));
```
Comment (maximecb) [inline yjit/src/codegen.rs:5590]: 'argv' usually means "argument value". Could probably use a more descriptive variable name.

File: yjit/src/codegen.rs [hunk 3]
```diff
@@ -5557,8 +5557,54 @@ fn gen_send_general(
...
+                            recv_load,
+                            EC,
+                            argc.into(),
+                            argv,
+                            kw_splat.into(),
+                            VM_BLOCK_HANDLER_NONE.into(),
+                        ]);
+
+                        ctx.stack_pop(argc as usize + 1);
+
```
Comment (maximecb) [inline yjit/src/codegen.rs:5602]:
```suggestion
```

File: yjit.c (hunk 1 above)
Comment (maximecb) [inline yjit.c:720]: Nitpick but should probably rename this to optimized call, as in optimized method.

File: yjit/src/stats.rs [hunk 4]
```diff
@@ -173,6 +173,11 @@ make_counters! {
     send_optimized_method,
     send_optimized_method_call,
     send_optimized_method_block_call,
+    send_call_block,
+    send_call_kwarg,
+    send_call_not_iseq,
+    send_call_lambda,
```
Comment (k0kubun) [inline yjit/src/stats.rs:179]:
```suggestion
```

Are these used somewhere?

**Review by maximecb — COMMENTED** [review]
Minor comments but this looks good. The implementation is pretty simple and I think deferring to a C call here is appropriate 👍

**Comment by jimmyhmiller** [comment]
@maximecb Glad to hear it :) Addressed all the feedback. So once it passes, should be all good.

**Review by maximecb — APPROVED** [review]

File: yjit.c [hunk 5]
```diff
@@ -716,6 +716,15 @@ rb_get_iseq_body_param_opt_table(const rb_iseq_t *iseq)
     return iseq->body->param.opt_table;
 }
 
+VALUE
+rb_optimized_call(VALUE *recv, rb_execution_context_t *ec, int argc, VALUE *argv, int kw_splat, VALUE block_handler)
```
Comment (nobu) [inline yjit.c:720]: Shouldn’t `*recv` be `recv`?

**Pull request merged**